### DISCUSSION TITLE
Asymmetry between SwarmEvent::ConnectionEstablished and SwarmEvent::ConnectionClosed

### DISCUSSION THREAD

**Discussion by nazar-pc** [libp2p/rust-libp2p#3418] [Q&A]
Just had a bug in the app due to `SwarmEvent::ConnectionClosed` happening much more often than `SwarmEvent::ConnectionEstablished`. I guess this happens due to connection errors, but I'm still surprised that connections are closing before being established.

Any suggestions how to account for this properly?

**Comment by mxinden** [comment]
I am surprised that this is the case. I would expect the number of `ConnectionClosed` and `ConnectionEstablished` events to only differ by the number of currently established connections. Can you provide a way to reproduce this @nazar-pc?

**Reply by nazar-pc** [reply]
I don't have a reliable reproduction, but it did happen numerous times under certain conditions (it was non-deterministic), I had underflow of the unsigned integer that was supposed to store number of established connections.

**Reply by mxinden** [reply]
Are you using `Swarm::ban_peer_id` @nazar-pc?

**Reply by nazar-pc** [reply]
Yes and started using relatively recently

**Reply by mxinden** [reply]
@nazar-pc can you still reproduce these failures with latest master or the `v0.51.0` release? The banning logic and the connection state reporting has been refactored with https://github.com/libp2p/rust-libp2p/pull/3328, thus I wonder whether the issue you are seeing is still occurring.

**Reply by nazar-pc** [reply]
I added some additional loud logging in case this happens and didn't see anything recently, so maybe it is fixed indeed (running quite recent `master` branch). Will keep an eye on it, thanks!

**Reply by mxinden** [reply]
Thanks @nazar-pc!
